rfcs/2022-02-11-responsive-and-adaptive-layout-strategy.md

```diff
@@ -0,0 +1,394 @@
+
+    ```jsx
+    <Text fontSize="100" />
+
+    <Text fontSize={{ small: 100, medium: 100, large: 100, xl: 400 }} />
+    ```
+
+- Hence the idea here is that all these components except the Layout components will handle responsiveness based on predefined breakpoints internally without the consumers explicitly defining anything for them.
+
+### 2. Components that can give some flexibility to handle Responsiveness to consumers
+
+- While most of the components can handle responsiveness internally there are use cases for components where we would need to give flexibility to accept responsive values for certain props. For eg: Layout Components that handles the overall layout of the page might change the layout on different screen sizes based on the available space.
+  <img alt="Adaptive Layout with flex direction" src="./images/responsive-adaptive-layout/adaptive-layout-with-flex-direction.png">
+- The idea here is to have certain components that deal with Layout of the page expose some flexibility to consumers to tweak if they want the layout to look different on different screen sizes
+
+### 3. Components for Adaptive layout(rendering components conditionally)
+
+- After handling responsiveness internally and also giving flexibility for few scenarios there are still few complex cases that we need to handle:
+
+  - There are cases where we might need to render a completely different variation of a component based on the screen size and user experience. For eg: A Search Field component might be different for larger screen sizes and smaller screen sizes.
+    <img alt="Responsive Modal" src="./images/responsive-adaptive-layout/adaptive-search-field.png">
+
+  - Another example is to show and hide certain elements based on the screen size so that we can optimise for user experience for each screen size
+    <img alt="Responsive Modal" src="./images/responsive-adaptive-layout/adaptive-layout-with-multiple-layouts.png">
+
+- The idea here is to make it feasible for consumers to handle these kind of complex use cases with ease, maybe with a set of predefined components and APIs.
+
+## How will we actually implement the approaches discussed above?
+
+### On Blade Code Side
+
+- To begin with, as a first step we will first need to setup different screen sizes that we'll support and store it in tokens. This will handle our layout changes
+
+  ```
+  // confirm with Saurav once
+  s: 320
+  m: 480
+  l: 760
+  xl: 960
+  2xl: 1200
+  3xl: 1600
+  ```
+
+- We would then need to add a listener in the `BladeProvider` which will listen to screen size changes and trigger a layout change.
+
+- For handling typography for different screen sizes we already have 2 scales one for smaller screens which are less than `960px` and the other for larger screens which are greater than `960px`. Now, once the screen size changes, the above listener would trigger a state change and then we can listen to it and swap the scales between small and large screens.
```

Comment:
This sounds like it may "stutter / freeze" the UI if someone switches between different resolutions (for example decrease browser width) (compared to responsive layout with media queries since now transitions would be more resource intensive) 🤔

Reply:
Won't media queries be used for standard CSS changes for responsive components? 

I feel the listener would be useful for adaptive components. 

That is how we have solved this in docs. We use [getBreakpointStyles](https://github.com/razorpay/docs/blob/master/docs-website/src/components/utils/breakpoints.ts#L45) function to add media queries to each responsive components, like [here for setting responsive padding](https://github.com/razorpay/docs/blob/master/docs-website/src/home/Payments/Payments.tsx#L25)

Whereas we use [useBreakpoint](https://github.com/razorpay/docs/blob/master/docs-website/src/components/utils/useBreakpoint.ts) for adaptive components

Reply:
> This sounds like it may "stutter / freeze" the UI if someone switches between different resolutions (for example decrease browser width) (compared to responsive layout with media queries since now transitions would be more resource intensive) 🤔

Why will it stutter? It's same as responsive only diff is we will switch things based on the defined breakpoints and not on every resize events

Reply:
What is the advantage of doing this using JS rather than using media queries for responsive components?